core/src/main/java/com/linecorp/armeria/server/Http2ServerUpgradeCodec.java

```diff
@@ -0,0 +1,166 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.linecorp.armeria.server;
+
+import static io.netty.handler.codec.base64.Base64Dialect.URL_SAFE;
+import static io.netty.handler.codec.http2.Http2CodecUtil.FRAME_HEADER_LENGTH;
+import static io.netty.handler.codec.http2.Http2CodecUtil.HTTP_UPGRADE_SETTINGS_HEADER;
+import static io.netty.handler.codec.http2.Http2CodecUtil.writeFrameHeader;
+import static io.netty.handler.codec.http2.Http2FrameTypes.SETTINGS;
+
+import java.nio.CharBuffer;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.base64.Base64;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http2.DefaultHttp2FrameReader;
+import io.netty.handler.codec.http2.Http2ConnectionHandler;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.handler.codec.http2.Http2Flags;
+import io.netty.handler.codec.http2.Http2FrameAdapter;
+import io.netty.handler.codec.http2.Http2FrameReader;
+import io.netty.handler.codec.http2.Http2Settings;
+import io.netty.util.CharsetUtil;
+
+/**
+ * Server-side codec for performing a cleartext upgrade from HTTP/1.x to HTTP/2.
+ */
+final class Http2ServerUpgradeCodec implements HttpServerUpgradeHandler.UpgradeCodec {
+
+    // Forked from http://github.com/netty/netty/blob/cf624c93c5f97097f1b13fe926ed50c32c8b1430/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerUpgradeCodec.java
+
+    private static final Logger logger = LoggerFactory.getLogger(Http2ServerUpgradeCodec.class);
+
+    private final Http2ConnectionHandler connectionHandler;
+    private final Http2FrameReader frameReader;
+
+    @Nullable
+    private Http2Settings settings;
+
+    /**
+     * Creates the codec using a default name for the connection handler when adding to the
+     * pipeline.
+     *
+     * @param connectionHandler the HTTP/2 connection handler
+     */
+    Http2ServerUpgradeCodec(Http2ConnectionHandler connectionHandler) {
+        this.connectionHandler = connectionHandler;
+        frameReader = new DefaultHttp2FrameReader();
+    }
+
+    @Override
+    public boolean prepareUpgradeResponse(ChannelHandlerContext ctx, HttpRequest upgradeRequest) {
+        try {
+            // Decode the HTTP2-Settings header and set the settings on the handler to make
+            // sure everything is fine with the request.
+            final List<String> upgradeHeaders = upgradeRequest.headers().getAll(HTTP_UPGRADE_SETTINGS_HEADER);
+            if (upgradeHeaders.size() != 1) {
+                throw new IllegalArgumentException(
+                        "There must be 1 and only 1 " + HTTP_UPGRADE_SETTINGS_HEADER + " header.");
```

Comment:
Should we send a 400 Bad Request response, handled by `ServerErrorHandler` if possible, instead? Can we add a test case for this?